src/loki/search.cc

```diff
@@ -724,10 +716,15 @@ namespace loki {
 std::unordered_map<valhalla::baldr::Location, PathLocation>
 Search(const std::vector<valhalla::baldr::Location>& locations,
        GraphReader& reader,
-       const DynamicCost* costing) {
+       const std::shared_ptr<DynamicCost>& costing) {
```

Comment:
Now that `NoCosting` exists, could we make this parameter just `const DynamicCost &costing` ?  It would mean:

  1. The runtime check below isn't needed.
  2. Other places where it's *assumed* that `costing` is non-null would actually be safe.

Reply:
maybe but im not going to do that in this PR since its already huge. at the moment the cost factory returns a shared pointer. i agree that it seems like we might be able to avoid doing that and refactor a lot of the code base to match but i can tell you it will be a decently large change that doesnt make sense to be added to this pr.